File: openrndr-adopted-artifex/src/main/java/io/lacuna/artifex/utils/DoubleAccumulator.java
package io.lacuna.artifex.utils;

import static java.lang.System.arraycopy;

/**
 * @author ztellman
 */
public class DoubleAccumulator {

  private double[] values = new double[2];
  private int index = 0;

  private void expand() {
    double[] values = new double[this.values.length << 1];
    arraycopy(this.values, 0, values, 0, this.values.length);
    this.values = values;
  }

  public DoubleAccumulator add(double[] ns) {
    while (index > this.values.length - ns.length) {
      expand();
    }

    System.arraycopy(ns, 0, values, index, ns.length);
    index += ns.length;

    return this;
  }

  public DoubleAccumulator add(double n) {
    if (index > this.values.length - 1) {
      expand();
    }

    values[index++] = n;

    return this;
  }

  public DoubleAccumulator add(double a, double b) {
    if (index > this.values.length - 2) {
      expand();
    }

    values[index++] = a;
    values[index++] = b;

    return this;
  }

  public void pop(int num) {
    index -= num;
  }

  public void clear() {
    index = 0;
  }

  public int size() {
    return index;
  }

  public double get(int index) {
    return values[index];
  }

  public double last() {
    return values[index - 1];
  }

  public void set(int index, double n) {
    values[index] = n;
  }

  public double[] toArray() {
    double[] result = new double[size()];
    arraycopy(values, 0, result, 0, size());
    return result;
  }

  @Override
  public String toString() {
    StringBuilder b = new StringBuilder();
    b.append("[");
    for (int i = 0; i < index; i++) {
      if (i != 0) {
        b.append(", ");
      }
      b.append(values[i]);
    }
    b.append("]");

    return b.toString();
  }
}

File: openrndr-kartifex/src/commonMain/kotlin/utils/DoubleAccumulator.kt
package utils

class DoubleAccumulator {
    private var values = DoubleArray(2)
    private var index = 0
    private fun expand() {
        val values = DoubleArray(values.size shl 1)
        //System.arraycopy(this.values, 0, values, 0, this.values.size)
        this.values.copyInto(values, 0)

        this.values = values
    }

    fun add(ns: DoubleArray): DoubleAccumulator {
        while (index > values.size - ns.size) {
            expand()
        }
        //java.lang.System.arraycopy(ns, 0, values, index, ns.size)
        ns.copyInto(values, index, 0, ns.size)

        index += ns.size
        return this
    }

    fun add(n: Double): DoubleAccumulator {
        if (index > values.size - 1) {
            expand()
        }
        values[index++] = n
        return this
    }

    fun add(a: Double, b: Double): DoubleAccumulator {
        if (index > values.size - 2) {
            expand()
        }
        values[index++] = a
        values[index++] = b
        return this
    }

    fun pop(num: Int) {
        index -= num
    }

    fun clear() {
        index = 0
    }

    fun size(): Int {
        return index
    }

    operator fun get(index: Int): Double {
        return values[index]
    }

    fun last(): Double {
        return values[index - 1]
    }

    operator fun set(index: Int, n: Double) {
        values[index] = n
    }

    fun toArray(): DoubleArray {
        val result = DoubleArray(size())
        values.copyInto(result, 0, 0, size())
        return result
    }

    override fun toString(): String {
        val b = StringBuilder()
        b.append("[")
        for (i in 0 until index) {
            if (i != 0) {
                b.append(", ")
            }
            b.append(values[i])
        }
        b.append("]")
        return b.toString()
    }
}
